src/Cli/dotnet/Installer/Windows/Security/Signature.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.Versioning;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using Windows.Win32.Foundation;
+using Windows.Win32.Security.Cryptography;
+using Windows.Win32.Security.WinTrust;
+using static Windows.Win32.PInvoke;
+
+namespace Microsoft.DotNet.Installer.Windows.Security
+{
+    /// <summary>
+    /// Contains methods for verifying Authenticode signatures on Windows.
+    /// </summary>
+#if NETCOREAPP
+    [SupportedOSPlatform("windows5.1.2600")]
+#endif
+    internal static class Signature
+    {
+        /// <summary>
+        /// Verifies that the certificate used to sign the specified file terminates in a trusted Microsoft root certificate. The policy check is performed against
+        /// the first simple chain. 
+        /// </summary>
+        /// <param name="path">The path of the file to verify.</param>
+        /// <returns>0 if the policy could be checked.<see cref="Marshal.GetPInvokeErrorMessage(int)"/> can be called to obtain more detail about the failure.</returns>
+        /// <exception cref="CryptographicException"/>
+        /// <remarks>This method does not perform any other chain validation like revocation checks, timestamping, etc.</remarks>
+        internal static unsafe int HasMicrosoftTrustedRoot(string path)
+        {
+            // Create an X509Certificate2 instance so we can access the certificate context and create a chain context.
+            using X509Certificate2 certificate = new(path);
```

Comment:
I feel like WinVerifyTrust already exposes back the certificate, and this has to replicate that work.  I advise checking to see if you can already get this from `trustData->pCert`.  If so, then between the action call to WinVerifyTrust and the cleanup call, do `PCCERT_CONTEXT pCertContext = CertDuplicateCertificateContext(trustData->pCert->psCertContext);`, and pass the PCCERT_CONTEXT (or `CERT_CONTEXT*`, I suppose) into this method instead of `string path`.

It's both more performant, and would avoid a theoretical ToC/ToU problem if the file isn't locked during your call.

Reply:
I don't think I have access to it. Since I'm verifying a file, the union struct is a `WINTRUT_FILE_INFO*` and not a `WINTRUST_CERT_INFO*`. In this case, the WinVerifyTrust call creates the cert context from the file, but won't expose that back to me.

I suppose I could do it the other way around, extract the cert context from the file, then let `WinVerifyTrust` verify the cert context, but I'll need to check if the OS does anything else when asked to verify a file.

Reply:
Ah, I thought that some of the other data got populated by calling into WinVerifyTrust; and had glossed over it being a union type.

So, disregard.